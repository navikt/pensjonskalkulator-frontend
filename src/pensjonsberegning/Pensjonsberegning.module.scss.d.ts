import globalClassNames from '../style.d'
declare const classNames: typeof globalClassNames & {
  readonly container: 'container'
  readonly sammenligning: 'sammenligning'
}
export = classNames
