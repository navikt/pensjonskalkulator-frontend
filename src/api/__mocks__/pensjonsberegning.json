[
  {
    "pensjonsaar": 2024,
    "pensjonsbeloep": 300001,
    "alder": 67
  },
  {
    "pensjonsaar": 2025,
    "pensjonsbeloep": 300002,
    "alder": 68
  },
  {
    "pensjonsaar": 2026,
    "pensjonsbeloep": 300003,
    "alder": 69
  }
]
