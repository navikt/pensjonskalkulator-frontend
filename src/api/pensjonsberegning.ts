import { Pensjonsberegning } from '../Pensjonsberegning'
import { isPensjonsberegning } from './typeguards'

export const fetchPensjonsberegning = (): Promise<Pensjonsberegning[]> => {
  return fetch('/pensjon/kalkulator/api/pensjonsberegning', {
    method: 'GET',
  })
    .then((response) => {
      if (response.ok) {
        return response.json()
      }
      throw new Error('Something went wrong')
    })
    .then((data) => {
      if (!data.some(isPensjonsberegning)) {
        throw new Error(`Mottok ugyldig pensjonsberegning fra backend: ${data}`)
      }
      return data
    })
}
