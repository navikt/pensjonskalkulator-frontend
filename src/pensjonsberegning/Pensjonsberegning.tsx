import { useEffect, useState } from 'react'

import { Alert, BodyLong, Loader } from '@navikt/ds-react'

import { fetchPensjonsberegning } from '@/api/pensjonsberegning'

import styles from './Pensjonsberegning.module.scss'

const usePensjonsberegning = (): FetchedData<PensjonsberegningResponse[]> => {
  const [beregning, setBeregning] = useState<
    FetchedData<PensjonsberegningResponse[]>
  >({
    data: null,
    isLoading: true,
    hasError: false,
  })

  useEffect(() => {
    fetchPensjonsberegning()
      .then((data: PensjonsberegningResponse[]) => {
        setBeregning({
          data: data.sort((a, b) => a.pensjonsaar - b.pensjonsaar),
          isLoading: false,
          hasError: false,
        })
      })
      .catch((error) => {
        // TODO add error loggingto a server?
        console.warn(error)
        setBeregning({
          data: null,
          isLoading: false,
          hasError: true,
        })
      })
  }, [])

  return beregning
}

export function Pensjonsberegning() {
  const beregning = usePensjonsberegning()

  if (beregning.isLoading) {
    return <Loader data-testid="loader" />
  } else if (beregning.hasError || beregning.data.length === 0) {
    return (
      <Alert variant="error">
        Vi klarte ikke å kalkulere pensjonen din. Prøv igjen senere.
      </Alert>
    )
  }

  return (
    <section className={styles.sammenligning}>
      <BodyLong>
        Hvis du fortsetter å ha samme inntekt som du har i dag kan du tidligst
        gå av med pensjon ved <b>{beregning.data[0].alder} år</b> hvor du vil få
        utbetalt <b>{beregning.data[0].pensjonsbeloep} kroner</b> i året
      </BodyLong>
      {beregning.data[1] && (
        <BodyLong>
          Dersom du jobber frem til du er <b>{beregning.data[1].alder} år</b>,
          vil du få <b>{beregning.data[1].pensjonsbeloep} kroner</b> utbetalt
        </BodyLong>
      )}
      {beregning.data[2] && (
        <BodyLong>
          Dersom du jobber frem til du er <b>{beregning.data[2].alder} år</b>,
          vil du få <b>{beregning.data[2].pensjonsbeloep} kroner</b> utbetalt
        </BodyLong>
      )}
    </section>
  )
}
