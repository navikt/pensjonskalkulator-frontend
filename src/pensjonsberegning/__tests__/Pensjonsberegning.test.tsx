import { render, screen, waitFor } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'

import { Pensjonsberegning } from '@/pensjonsberegning'
import {
  createFailureFetchResponse,
  createSuccessFetchResponse,
} from '@/test-utils'

const cachedFetch = global.fetch

const data: Pensjonsberegning[] = [
  {
    pensjonsbeloep: 1234,
    pensjonsaar: 2020,
    alder: 67,
  },
  {
    pensjonsbeloep: 1357,
    pensjonsaar: 2021,
    alder: 68,
  },
  {
    pensjonsbeloep: 2345,
    pensjonsaar: 2023,
    alder: 70,
  },
]

function mockSuccessResponse() {
  const fetchMock = vi.fn().mockResolvedValue(createSuccessFetchResponse(data))
  global.fetch = fetchMock
}

function mockErrorResponse() {
  const fetchMock = vi.fn().mockRejectedValue(createFailureFetchResponse(500))
  global.fetch = fetchMock
}

function cleanup() {
  global.fetch = cachedFetch
}

describe('Pensjonsberegning', () => {
  afterEach(() => {
    cleanup()
  })

  it('viser pensjonsberegning hentet fra backend', async () => {
    mockSuccessResponse()
    render(<Pensjonsberegning />)

    await waitFor(() => {
      expect(screen.getByText(`${data[0].alder} år`)).toBeVisible()
      expect(screen.getByText(`${data[1].alder} år`)).toBeVisible()
      expect(screen.getByText(`${data[2].alder} år`)).toBeVisible()
      expect(screen.getByText(`${data[0].pensjonsbeloep} kroner`)).toBeVisible()
      expect(screen.getByText(`${data[1].pensjonsbeloep} kroner`)).toBeVisible()
      expect(screen.getByText(`${data[2].pensjonsbeloep} kroner`)).toBeVisible()
    })
  })

  it('viser loader før pensjonsberegningen har blitt hentet', () => {
    mockSuccessResponse()
    render(<Pensjonsberegning />)

    expect(screen.getByTestId('loader')).toBeVisible()
  })

  it('viser feilmelding om henting av pensjonberegning feiler', async () => {
    mockErrorResponse()

    render(<Pensjonsberegning />)

    await waitFor(() => {
      expect(
        screen.getByText(
          'Vi klarte ikke å kalkulere pensjonen din. Prøv igjen senere.'
        )
      ).toBeVisible()
    })
  })
})
