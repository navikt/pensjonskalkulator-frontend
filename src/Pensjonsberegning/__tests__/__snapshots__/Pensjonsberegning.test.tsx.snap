// Vitest Snapshot v1

exports[`Pensjonsberegning > viser feilmelding om henting av pensjonberegning feiler 1`] = `
<DocumentFragment>
  <div
    class="navds-alert navds-alert--error navds-alert--medium"
  >
    <svg
      aria-labelledby="title-r2"
      class="navds-alert__icon"
      fill="none"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="title-r2"
      >
        Feil
      </title>
      <path
        clip-rule="evenodd"
        d="M0 11.976C.013 5.372 5.395 0 11.999 0 18.639.013 24.012 5.406 24 12.022 23.986 18.627 18.603 24 12 24h-.023a11.924 11.924 0 0 1-8.48-3.531A11.923 11.923 0 0 1 0 11.976Zm17-3.547L15.571 7 12 10.572 8.429 7 7 8.429 10.572 12 7 15.571 8.429 17 12 13.428 15.571 17 17 15.571 13.428 12 17 8.429Z"
        fill="#C30000"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="navds-alert__wrapper navds-body-long"
    >
      <h1
        class="navds-heading navds-heading--small navds-typo--spacing"
      >
        Vi klarte ikke å kalkulere pensjonen din. Prøv igjen senere.
      </h1>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Pensjonsberegning > viser feilmelding om henting av pensjonberegning feiler 2`] = `
<DocumentFragment>
  <div
    class="navds-alert navds-alert--error navds-alert--medium"
  >
    <svg
      aria-labelledby="title-r3"
      class="navds-alert__icon"
      fill="none"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="title-r3"
      >
        Feil
      </title>
      <path
        clip-rule="evenodd"
        d="M0 11.976C.013 5.372 5.395 0 11.999 0 18.639.013 24.012 5.406 24 12.022 23.986 18.627 18.603 24 12 24h-.023a11.924 11.924 0 0 1-8.48-3.531A11.923 11.923 0 0 1 0 11.976Zm17-3.547L15.571 7 12 10.572 8.429 7 7 8.429 10.572 12 7 15.571 8.429 17 12 13.428 15.571 17 17 15.571 13.428 12 17 8.429Z"
        fill="#C30000"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="navds-alert__wrapper navds-body-long"
    >
      <h1
        class="navds-heading navds-heading--small navds-typo--spacing"
      >
        Vi klarte ikke å kalkulere pensjonen din. Prøv igjen senere.
      </h1>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Pensjonsberegning > viser loading og deretter pensjonsberegning hentet fra backend 1`] = `
<DocumentFragment>
  <section
    class="_sammenligning_e128d6"
  >
    <p
      class="navds-body-long"
    >
      Hvis du fortsetter å ha samme inntekt som du har i dag kan du tidligst gå av med pensjon ved 
      <b>
        67 år
      </b>
       hvor du vil få utbetalt 
      <b>
        300001 kroner
      </b>
       i året
    </p>
    <p
      class="navds-body-long"
    >
      Dersom du jobber frem til du er 
      <b>
        68 år
      </b>
      , vil du få 
      <b>
        300002 kroner
      </b>
       utbetalt
    </p>
    <p
      class="navds-body-long"
    >
      Dersom du jobber frem til du er 
      <b>
        69 år
      </b>
      , vil du få 
      <b>
        300003 kroner
      </b>
       utbetalt
    </p>
  </section>
</DocumentFragment>
`;
